File: Tests/BuilderUIAuditTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

@MainActor
final class BuilderUIAuditTests: XCTestCase {
    func testLegacyTargetsBecomeEditableWithoutChangingExecutionOrOtherDraftValues() {
        let inherited = SetPlan(grip: GripSpec(), repsPerSide: 0)
        let kilograms = SetPlan(grip: GripSpec(), targetLoKg: 6, targetHiKg: 9)
        let percentage = SetPlan(grip: GripSpec(), targetLoPercent: 0.4, targetHiPercent: 0.5)
        var original = RoutineDraft()
        original.plan.name = "  Unfinished  "
        original.plan.sets = [inherited, kilograms, percentage]
        original.plan.targetLoPercent = 0.2
        original.plan.targetHiPercent = 0.3
        let edited = BuilderDraftPreparation.editable(original)
        XCTAssertEqual(edited.plan.name, original.plan.name)
        XCTAssertEqual(edited.plan.sets[0].repsPerSide, 0)
        XCTAssertEqual(edited.plan.sets[1], kilograms)
        XCTAssertEqual(edited.plan.sets[2], percentage)
        XCTAssertNil(edited.plan.targetPercentBand)
        XCTAssertEqual(original.plan.targetPercentBand, 0.2...0.3)
        XCTAssertEqual(BuilderDraftPreparation.editable(edited), edited)
        for maxKg in [nil, 0.5, 20.0, 60.0] as [Double?] {
            for index in original.plan.sets.indices {
                XCTAssertEqual(PlanMath.targetBand(edited.plan.sets[index], in: edited.plan, maxKg: maxKg),
                    PlanMath.targetBand(original.plan.sets[index], in: original.plan, maxKg: maxKg))
            }
        }
    }

    func testClearingMaterializedTargetDoesNotRevealAnInvisibleRoutineFallback() {
        var original = RoutineDraft()
        original.plan.sets = [SetPlan(grip: GripSpec())]
        original.plan.targetLoPercent = 0.2
        original.plan.targetHiPercent = 0.3
        var edited = BuilderDraftPreparation.editable(original)
        edited.plan.sets[0].targetLoPercent = nil
        edited.plan.sets[0].targetHiPercent = nil
        XCTAssertNil(PlanMath.targetBand(edited.plan.sets[0], in: edited.plan, maxKg: 30))
        XCTAssertEqual(BuilderDraftPreparation.editable(.blank(named: "")), .blank(named: ""))
    }

    func testBandDraggingPreservesWidthAtBothWallsAndFitsOversizedIncomingValues() {
        XCTAssertEqual(BandTrimmerMath.translated(lower: 5, upper: 10, delta: 100, scale: 0...20, step: 0.5), 15...20)
        XCTAssertEqual(BandTrimmerMath.translated(lower: 5, upper: 10, delta: -100, scale: 0...20, step: 0.5), 0...5)
        XCTAssertEqual(BandTrimmerMath.translated(lower: -5, upper: 30, delta: 1, scale: 0...20, step: 0.5), 0...20)
        XCTAssertEqual(BandTrimmerMath.translated(lower: 0, upper: 0, delta: 1, scale: 0...0, step: 0.5), 0...0)
        XCTAssertEqual(BandTrimmerMath.translated(lower: 1, upper: 2, delta: 1, scale: 1...100, step: 5), 1...2)
    }

    func testConsistencyHintDoesNotPretendALapsedUserIsNew() {
        func record(_ tracked: Bool, completed: Int = 0) -> DayRecord {
            DayRecord(day: DayStamp(raw: 20_000), completed: completed, target: 2, tracked: tracked, climb: nil)
        }
        XCTAssertTrue(ConsistencyEmptyState.showsFirstUseHint([record(false), record(true)]))
        XCTAssertFalse(ConsistencyEmptyState.showsFirstUseHint([record(true), record(true)]))
        XCTAssertFalse(ConsistencyEmptyState.showsFirstUseHint([record(false), record(true, completed: 1)]))
        XCTAssertFalse(ConsistencyEmptyState.showsFirstUseHint([]))
    }

    func testBatteryPercentageUsesOneBoundedRoundingRule() {
        XCTAssertEqual(BatteryDisplay.percentage(0.556), 56)
        XCTAssertEqual(BatteryDisplay.percentage(0.554), 55)
        XCTAssertEqual(BatteryDisplay.percentage(-0.1), 0)
        XCTAssertEqual(BatteryDisplay.percentage(1.2), 100)
        XCTAssertEqual(BatteryDisplay.percentage(.nan), 0)
    }
}

File: android/app/src/test/kotlin/run/nuri/getagrip/ui/BuilderUIAuditTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.cancel
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.test.advanceTimeBy
import kotlinx.coroutines.test.runCurrent
import run.nuri.getagrip.ui.builder.DraftStashCoalescer
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue
import run.nuri.getagrip.engine.*
import run.nuri.getagrip.ui.builder.BuilderDraft
import run.nuri.getagrip.ui.components.BandTrimmerMath
import run.nuri.getagrip.ui.components.BatteryDisplay
import run.nuri.getagrip.ui.today.ConsistencyEmptyState

@OptIn(ExperimentalCoroutinesApi::class)
class BuilderUIAuditTests {
    @Test fun heldInputSavesTheLatestDraftEveryWindowAndCancelsWithTheBuilder() = runTest {
        val owner = CoroutineScope(coroutineContext + Job())
        val written = mutableListOf<Int>()
        var draft = 1
        val stash = DraftStashCoalescer(owner) { written += draft }
        stash.changed()
        runCurrent()
        repeat(4) {
            advanceTimeBy(100)
            draft += 1
            stash.changed()
        }
        advanceTimeBy(100)
        runCurrent()
        assertEquals(listOf(5), written)
        draft = 6
        stash.changed()
        runCurrent()
        advanceTimeBy(500)
        runCurrent()
        assertEquals(listOf(5, 6), written)
        draft = 7
        stash.changed()
        runCurrent()
        owner.cancel()
        advanceTimeBy(500)
        runCurrent()
        assertEquals(listOf(5, 6), written)
    }

    @Test fun legacyTargetsBecomeEditableWithoutChangingExecutionOrOtherDraftValues() {
        val inherited = SetPlan(grip = GripSpec(), repsPerSide = 0)
        val kilograms = SetPlan(grip = GripSpec(), targetLoKg = 6.0, targetHiKg = 9.0)
        val percentage = SetPlan(grip = GripSpec(), targetLoPercent = 0.4, targetHiPercent = 0.5)
        val original = RoutineDraft(plan = SessionPlan(name = "  Unfinished  ",
            sets = listOf(inherited, kilograms, percentage), targetLoPercent = 0.2, targetHiPercent = 0.3))
        val edited = BuilderDraft.editable(original)
        assertEquals(original.plan.name, edited.plan.name)
        assertEquals(0, edited.plan.sets[0].repsPerSide)
        assertEquals(kilograms, edited.plan.sets[1])
        assertEquals(percentage, edited.plan.sets[2])
        assertNull(edited.plan.targetPercentBand)
        assertEquals(0.2..0.3, original.plan.targetPercentBand)
        assertEquals(edited, BuilderDraft.editable(edited))
        for (maxKg in listOf(null, 0.5, 20.0, 60.0)) {
            for (index in original.plan.sets.indices) {
                assertEquals(PlanMath.targetBand(original.plan.sets[index], original.plan, maxKg),
                    PlanMath.targetBand(edited.plan.sets[index], edited.plan, maxKg))
            }
        }
    }

    @Test fun clearingMaterializedTargetDoesNotRevealAnInvisibleRoutineFallback() {
        val original = RoutineDraft(plan = SessionPlan(sets = listOf(SetPlan(grip = GripSpec())),
            targetLoPercent = 0.2, targetHiPercent = 0.3))
        val edited = BuilderDraft.editable(original)
        val cleared = edited.plan.sets[0].copy(targetLoPercent = null, targetHiPercent = null)
        assertNull(PlanMath.targetBand(cleared, edited.plan, 30.0))
        val blank = RoutineDraft.blank("")
        assertEquals(blank, BuilderDraft.editable(blank))
    }

    @Test fun bandDraggingPreservesWidthAtBothWallsAndFitsOversizedIncomingValues() {
        assertEquals(15.0..20.0, BandTrimmerMath.translated(5.0, 10.0, 100.0, 0.0..20.0, 0.5))
        assertEquals(0.0..5.0, BandTrimmerMath.translated(5.0, 10.0, -100.0, 0.0..20.0, 0.5))
        assertEquals(0.0..20.0, BandTrimmerMath.translated(-5.0, 30.0, 1.0, 0.0..20.0, 0.5))
        assertEquals(0.0..0.0, BandTrimmerMath.translated(0.0, 0.0, 1.0, 0.0..0.0, 0.5))
        assertEquals(1.0..2.0, BandTrimmerMath.translated(1.0, 2.0, 1.0, 1.0..100.0, 5.0))
    }

    @Test fun consistencyHintDoesNotPretendALapsedUserIsNew() {
        fun record(tracked: Boolean, completed: Int = 0) =
            DayRecord(DayStamp(20_000), completed, 2, tracked, null)
        assertTrue(ConsistencyEmptyState.showsFirstUseHint(listOf(record(false), record(true))))
        assertFalse(ConsistencyEmptyState.showsFirstUseHint(listOf(record(true), record(true))))
        assertFalse(ConsistencyEmptyState.showsFirstUseHint(listOf(record(false), record(true, 1))))
        assertFalse(ConsistencyEmptyState.showsFirstUseHint(emptyList()))
    }

    @Test fun batteryPercentageUsesOneBoundedRoundingRule() {
        assertEquals(56, BatteryDisplay.percentage(0.556))
        assertEquals(55, BatteryDisplay.percentage(0.554))
        assertEquals(0, BatteryDisplay.percentage(-0.1))
        assertEquals(100, BatteryDisplay.percentage(1.2))
        assertEquals(0, BatteryDisplay.percentage(Double.NaN))
    }
}
